File: src/main/java/com/bajdcc/util/lexer/automata/dfa/DFAEdgeBag.java
package com.bajdcc.util.lexer.automata.dfa;

import com.bajdcc.util.lexer.automata.nfa.NFAStatus;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * DFA边辅助数据结构
 *
 * @author bajdcc
 */
public class DFAEdgeBag extends DFAEdgeData {
	/**
	 * NFA状态集合
	 */
	public Set<NFAStatus> nfaStatus = new HashSet<>();

	/**
	 * 获得状态编号描述（逗号分隔）
	 *
	 * @param dfaStatusList 状态表
	 * @return 状态编号描述
	 */
	public String getStatusString(List<NFAStatus> dfaStatusList) {
		StringBuilder sb = new StringBuilder();
		for (NFAStatus status : nfaStatus) {
			sb.append(dfaStatusList.indexOf(status)).append(",");
		}
		return sb.toString();
	}
}

File: src/main/java/com/bajdcc/util/lexer/automata/dfa/DFAEdgeBag.kt
package com.bajdcc.util.lexer.automata.dfa

import com.bajdcc.util.lexer.automata.EdgeType
import com.bajdcc.util.lexer.automata.nfa.NFAEdge
import com.bajdcc.util.lexer.automata.nfa.NFAStatus

/**
 * DFA边辅助数据结构
 * [type] 边类型
 * [param] 数据
 * [nfaEdges] NFA边集合
 * [nfaStatus] NFA状态集合
 * @author bajdcc
 */
data class DFAEdgeBag(var type: EdgeType = EdgeType.EPSILON,
                      var param: Int = -1,
                      var nfaEdges: MutableList<NFAEdge> = mutableListOf(),
                      var nfaStatus: MutableSet<NFAStatus> = mutableSetOf()) {

    /**
     * 获得状态编号描述（逗号分隔）
     *
     * @param dfaStatusList 状态表
     * @return 状态编号描述
     */
    fun getStatusString(dfaStatusList: List<NFAStatus>): String {
        val sb = StringBuilder()
        for (status in nfaStatus) {
            sb.append(dfaStatusList.indexOf(status)).append(",")
        }
        return sb.toString()
    }
}
